xdsl/transforms/convert_ptr_to_llvm.py

```diff
@@ -0,0 +1,274 @@
+        if not any(isinstance(arg.type, ptr.PtrType) for arg in op.arguments):
+            return
+
+        insert_point = InsertPoint.before(op)
+        new_arguments: list[SSAValue] = []
+
+        # insert `ptr_xdsl.ptr -> llvm.ptr` casts for ptr return values
+        for argument in op.arguments:
+            if isinstance(argument.type, ptr.PtrType):
+                rewriter.insert_op(
+                    cast_op := builtin.UnrealizedConversionCastOp.get(
+                        [argument], [llvm.LLVMPointerType.opaque()]
+                    ),
+                    insert_point,
+                )
+                new_arguments.append(cast_op.results[0])
+            else:
+                new_arguments.append(argument)
+
+        rewriter.replace_matched_op(func.ReturnOp(*new_arguments))
+
+
+@dataclass
+class ConvertCallOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.CallOp, rewriter: PatternRewriter, /):
+        if not any(
+            isinstance(arg.type, ptr.PtrType) for arg in op.arguments
+        ) and not any(isinstance(type, ptr.PtrType) for type in op.result_types):
+            return
+
+        # rewrite arguments
+        insert_point = InsertPoint.before(op)
+        new_arguments: list[SSAValue] = []
+
+        # insert `ptr_xdsl.ptr -> llvm.ptr` casts for argument values
+        for argument in op.arguments:
+            if isinstance(argument.type, ptr.PtrType):
+                rewriter.insert_op(
+                    cast_op := builtin.UnrealizedConversionCastOp.get(
+                        [argument], [llvm.LLVMPointerType.opaque()]
+                    ),
+                    insert_point,
+                )
+                new_arguments.append(cast_op.results[0])
+            else:
+                new_arguments.append(argument)
+
+        insert_point = InsertPoint.after(op)
+        new_results: list[SSAValue] = []
+
+        #  insert `llvm.ptr -> ptr_xdsl.ptr` casts for return values
+        for result in op.results:
+            if isinstance(result.type, ptr.PtrType):
+                rewriter.insert_op(
+                    cast_op := builtin.UnrealizedConversionCastOp.get(
+                        [result],
+                        [result.type],
+                    ),
+                    insert_point,
+                )
+                new_results.append(cast_op.results[0])
+            else:
+                new_results.append(result)
+
+        new_return_types = [
+            llvm.LLVMPointerType.opaque() if isinstance(type, ptr.PtrType) else type
+            for type in op.result_types
+        ]
+
+        rewriter.replace_matched_op(
+            func.CallOp(op.callee, new_arguments, new_return_types)
+        )
+
+
+@dataclass
+class ConvertPtrAddOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: ptr.PtrAddOp, rewriter: PatternRewriter, /):
+        # TODO: could be one cast?
```

Comment:
please remove all todos in this file, either by addressing them, or by filing issues, let's discuss this